system/Validation/Validation.php

```diff
@@ -108,6 +108,10 @@ public function __construct($config, RendererInterface $view)
      */
     public function run(?array $data = null, ?string $group = null, ?string $dbGroup = null): bool
     {
+        // If there are still validation errors for redirect_with_input request, remove them.
+        // See `getErrors()` method.
+        unset($_SESSION['_ci_validation_errors']);
```

Comment:
`isset($_SESSION, $_SESSION['_ci_validation_errors']` check may be needed to avoid race condition.

```suggestion
        if (isset($_SESSION, $_SESSION['_ci_validation_errors'])) {
            unset($_SESSION['_ci_validation_errors']);
        }
```

Reply:
@samsonasik Can you explain the race condition and why the `if` avoid it?

Reply:
That's just in case, dealing with session is sometime like dealing with cache, which possibly already removed on concurrent request.

Reply:
Even if it was already removed by a concurrent request, there is no error when unsetting a non-existing variable.

Reply:
it seems show the error if $_SESSION not exists ( probably happen with unset($_SESSION)), see

- https://3v4l.org/p5Okf ( $_SESSION never defined)
- https://3v4l.org/E9mWR ($_SESSION defined but unset next for whatever reason)

Reply:
I've confirmed the code in a controller. Yes, you're correct. `unset($_SESSION['_ci_validation_errors']);` shows Undefined variable $_SESSION.